components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,133 @@
+
+package mozilla.components.service.glean.storages
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.io.File
+import java.io.FilenameFilter
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.FileReader
+import java.io.IOException
+import java.io.OutputStreamWriter
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * TelemetryStorage implementation that stores pings as files on disk.
+ */
+internal open class PingStorageEngine(context: Context) {
+
+    private val logger: Logger = Logger("glean/PingStorageEngine")
+
+    @VisibleForTesting
+    internal val storageDirectory: File = File(context.applicationInfo.dataDir, STORAGE_DIRECTORY)
+
+    companion object {
+        private const val FILE_PATTERN = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+        private const val STORAGE_DIRECTORY = "glean/storage"
+    }
+
+    @Synchronized
+    fun store(pingType: String, pingPath: String, ping: String) {
+        val pingStorageDirectory = File(storageDirectory, pingType)
+        assertDirectory(pingStorageDirectory)
+
+        // Build the file path for the ping, using the UUID from the path for the file name
+        val pingFile = File(pingStorageDirectory, pingPath.substring(pingPath.lastIndexOf('/')))
+
+        var stream: FileOutputStream? = null
+
+        try {
+            stream = FileOutputStream(pingFile, true)
+
+            val writer = BufferedWriter(OutputStreamWriter(stream))
+            writer.write(pingPath)
+            writer.newLine()
+            writer.write(ping)
+            writer.newLine()
+            writer.flush()
+            writer.close()
+        } catch (e: IOException) {
+            logger.warn("IOException while writing event to disk", e)
+        } finally {
+            stream?.close()
+        }
+    }
+
+    fun process(pingType: String, callback: (String, String) -> Boolean): Boolean {
+        listPingFiles(pingType).forEach { file ->
+            val reader: FileReader? = null
+
+            try {
+                val bufferedReader = BufferedReader(FileReader(file))
+                val path = bufferedReader.readLine()
+                val serializedPing = bufferedReader.readLine()
+
+                val processed = serializedPing == null || callback(path, serializedPing)
+
+                if (processed) {
+                    file?.delete()
+                } else {
+                    // The callback couldn't process this file.
+                    return false
+                }
+            } catch (e: FileNotFoundException) {
+                // This shouldn't happen after we queried the directory.
+            } catch (e: IOException) {
+                // Something is not right.
```

Comment:
Let's log these errors.  Down the road, we may want to record metrics about them, too, so maybe add a TODO?

Reply:
On second thought, if this is failing, we're unlikely to get any metrics out of this client anyway, so maybe not worth recording.